### PR TITLE
Add Jsut asr recipe

### PR DESCRIPTION
**Pull request by Masao-Someki**
add jsut asr recipe

### CODE REVIEW

**Comment by kan-bayashi** [comment]
Great!
Let me check some points:
- In the case of Japanese, we cannot calculate WER as the same as JNAS. Only CER is OK.
- It seems that you re-use data prep script of `egs/jsut/tts1`. In the case of `tts1`, we use Hiaragana instead of original text, but in the case `asr1` we do not need to use hiragana, just use original text including Hiragana, Katakana, and Kanji.

**Comment by Masao-Someki** [comment]
I thought just using only hiraganas are better than mix of kanjis and katakanas..
I will fix it!

File: egs/jsut/asr1/run.sh [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+feat_dt_dir=${dumpdir}/${train_dev}; mkdir -p ${feat_dt_dir}
+feat_ev_dir=${dumpdir}/${recog_set}; mkdir -p ${feat_ev_dir}
+
+if [ ${stage} -le 1 ] && [ ${stop_stage} -ge 1 ]; then
+    ### Task dependent. You have to design training and dev sets by yourself.
+    ### But you can utilize Kaldi recipes in most cases
+    echo "stage 1: Feature Generation"
+    fbankdir=fbank
+    # Generate the fbank features; by default 80-dimensional fbanks with pitch on each frame
+	steps/make_fbank_pitch.sh --cmd "$train_cmd" \
```
Comment (sw005320) [inline egs/jsut/asr1/run.sh]: Please use regular indent.

File: egs/jsut/asr1/run.sh [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    compute-cmvn-stats scp:data/${train_set}/feats.scp data/${train_set}/cmvn.ark
+
+    # dump features for training
+    dump.sh --cmd "$train_cmd" --nj ${nj} --do_delta ${do_delta} \
+        data/${train_set}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/train ${feat_tr_dir}
+    dump.sh --cmd "$train_cmd" --nj 4 --do_delta ${do_delta} \
+        data/${train_dev}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/dev ${feat_dt_dir}
+
+    for rtask in ${recog_set}; do
+
```
Comment (sw005320) [inline egs/jsut/asr1/run.sh]: Do not insert redudant blank lines.

File: egs/jsut/asr1/run.sh [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    text2token.py -s 1 -n 1 data/${train_dev}/text | cut -f 2- -d" " \
+	   > ${lmdatadir}/valid.txt
+
+    # use only 1 gpu for train.
+    if [ ${ngpu} -gt 1 ]; then
+	echo "LM training does not support multi-gpu. Singlu gpu will be used."
+    fi
+    lmgpu=1
+    ${cuda_cmd} --gpu ${lmgpu} ${lmexpdir}/train.log \
+	    lm_train.py \
```
Comment (sw005320) [inline egs/jsut/asr1/run.sh]: The indent of these lines seems to be broken.

**Comment by sw005320** [comment]
Did you get the current result (~20% CER) with the kanji/hiragana/katakana characters? How many characters did you use? I think this is a cool result by considering the fact it is based on small data.

**Comment by Masao-Someki** [comment]
Yes, I got this result with kanji/hiragana/katakanas. 
For example, the text contains sentences like
```
BASIC5000_0022 むしろロン毛のほうが禿げやすいって聞いたぞ。
BASIC5000_0023 むこうに立っている女の子は、メアリーです。
BASIC5000_0024 システィナ礼拝堂は、１４７３年に、バティカン宮殿内に建立された、壮大な礼拝堂です。
BASIC5000_0025 テニスにもあるけど、４大大会って何。
```

**Comment by Masao-Someki** [comment]
Yes, I got this result with kanji/hiragana/katakanas. 
For example, the text contains sentences like
```
BASIC5000_0022 むしろロン毛のほうが禿げやすいって聞いたぞ。
BASIC5000_0023 むこうに立っている女の子は、メアリーです。
BASIC5000_0024 システィナ礼拝堂は、１４７３年に、バティカン宮殿内に建立された、壮大な礼拝堂です。
BASIC5000_0025 テニスにもあるけど、４大大会って何。
```

File: egs/jsut/asr1/conf/fbank.conf [hunk 4]
```diff
@@ -0,0 +1,2 @@
+--sample-frequency=48000
```
Comment (sw005320) [inline egs/jsut/asr1/conf/fbank.conf]: We usually downsample the data to 16000 for ASR. Many front-end configurations are based on  (optimized by)16000. Could you compare 16000 vs. 48000?

You can just add the sox downsample command in `wav.scp`.
See https://github.com/espnet/espnet/blob/f7e3dbb8f1e011c41973c4c77403edf8ac684126/egs/hkust/asr1/run.sh#L107
(This example is upsample, but your case is downsample).

File: egs/jsut/asr1/conf/pitch.conf [hunk 5]
```diff
@@ -0,0 +1 @@
+--sample-frequency=48000
```
Comment (sw005320) [inline egs/jsut/asr1/conf/pitch.conf]: same comment.

File: egs/jsut/asr1/run.sh [hunk 6]
```diff
@@ -0,0 +1,235 @@
+#!/bin/bash
+
+# Copyright 2017 Johns Hopkins University (Shinji Watanabe)
```
Comment (sw005320) [inline egs/jsut/asr1/run.sh]: change it to your name.

File: egs/jsut/asr1/run.sh [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+# root directory of db
+db_root=./downloads
+
+# exp tag
+tag="" # tag for managing experiments.
+
+. utils/parse_options.sh || exit 1;
+
+. ./path.sh
```
Comment (sw005320) [inline egs/jsut/asr1/run.sh]: I don't remember why we have `. ./path.sh` and `. ./cmd.sh` twice. I'll put it in the GitHub issue.

File: egs/jsut/asr1/conf/fbank.conf (hunk 4 above)
Comment (Masao-Someki) [reply egs/jsut/asr1/conf/fbank.conf]: OK!

File: egs/jsut/asr1/run.sh (hunk 7 above)
Comment (Masao-Someki) [reply egs/jsut/asr1/run.sh]: I'll delete the second path.sh and cmd.sh.

Comment (sw005320) [reply egs/jsut/asr1/run.sh]: You don't have to do it now.
It seems that all recipes have such duplications and maybe there are some reasons.

Comment (Masao-Someki) [reply egs/jsut/asr1/run.sh]: sure.

**Comment by sw005320** [comment]
Did you compare the result with 16k vs. 48k?

**Comment by Masao-Someki** [comment]
Sorry, now I got that result.
```
write a CER (or TER) result in exp/tr_no_dev_pytorch_train/decode_eval1_decode/result.txt
      | SPKR   | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
      | Sum/Avg|  250    5928 | 38.0   39.6   22.4    5.4   67.4  100.0 |
write a WER result in exp/tr_no_dev_pytorch_train/decode_eval1_decode/result.wrd.txt
       | SPKR   | # Snt # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
       | Sum/Avg|  250    250 |  0.0  100.0    0.0    0.0  100.0  100.0 |
```
I think something in the data preparation process is wrong..
I'll check what is happening.

**Comment by Masao-Someki** [comment]
sorry, the model I thought was learned completely was not trained well.
I'll continue training the model little more.

**Comment by sw005320** [comment]
No problem.

**Comment by Masao-Someki** [comment]
compared with 16k, it seems the results of 48k is better.

File: egs/jsut/asr1/conf/slurm.conf [hunk 8]
```diff
@@ -6,7 +6,7 @@ option mem=0          # Do not add anything to qsub_opts
 option num_threads=* --cpus-per-task $0 --ntasks-per-node=1
 option num_threads=1 --cpus-per-task 1  --ntasks-per-node=1 # Do not add anything to qsub_opts
 default gpu=0
-option gpu=0 -p gpu
```
Comment (sw005320) [inline egs/jsut/asr1/conf/slurm.conf]: Please do not push the local change.

**Comment by sw005320** [comment]
> compared with 16k, it seems the results of 48k is better.

This is very interesting. Maybe we'll use higher order (80dim) filter bank, which will smooth the higher frequency bins more compared with the 16k setup.
Could you also try the 48k transformer?
Also, could you modify your script so that you can easily switch from 16k to 48k from `run.sh`?

File: egs/jsut/asr1/local/make_pitch.sh [hunk 9]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+#This script is to adjust sampling rate of the wav files and config files.
+
+pitch=$1
+
+# fix fbank.conf
+sed -i -e "s/sample-frequency=[0-9]*/sample-frequency=$pitch/" ./conf/fbank.conf
```
Comment (sw005320) [inline egs/jsut/asr1/local/make_pitch.sh]: This is a little bit risky way. My suggestion
1) prepare `conf/fbank_16k.conf` and `conf/pitch_16k.conf`, and git add them.
2) remove `conf/fbank.conf` and `conf/pitch.conf` by git rm
3) this script generate `./conf/fbank.conf` from `./conf/fbank_16k.conf
```
sed -e "s/sample-frequency=[0-9]*/sample-frequency=$pitch/" ./conf/fbank_16k.conf > ./conf/fbank.conf
```

File: egs/jsut/asr1/local/make_pitch.sh [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+#This script is to adjust sampling rate of the wav files and config files.
+
+pitch=$1
```
Comment (sw005320) [inline egs/jsut/asr1/local/make_pitch.sh]: pitch -> sampling_rate

File: egs/jsut/asr1/run.sh [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+fi
+
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
+    ### Task dependent. You have to make data the following preparation part by yourself.
+    ### But you can utilize Kaldi recipes in most cases
+    echo "stage 0: Data Preparation"
+    # Initial normalization of the data
+    local/data_prep.sh ${db_root}/jsut_ver1.1 data/train $sampling_rate
+    utils/validate_data_dir.sh --no-feats data/train
+    # rebuilding the pitch.conf and fbank.conf
```
Comment (sw005320) [inline egs/jsut/asr1/run.sh]: `# changing the sampling rate option in pitch.conf and fbank.conf`

File: egs/jsut/asr1/local/make_pitch.sh (hunk 9 above)
Comment (Masao-Someki) [reply egs/jsut/asr1/local/make_pitch.sh]: Okay!

**Comment by sw005320** [comment]
Thanks. After you update the result with 48k transformer, I'll merge this PR.

**Comment by Masao-Someki** [comment]
I simply thought that with 48k dataset, transformer model could reach to a better result...
I don't know why rnn model(elayer=4, units=1024) with 48k dataset got the better result than transformer with 16k, and worse results with 48k-transformer compared to 16k-transformer.
@sw005320 Do you have any ideas..?

I will upload trained model later. I'm in a bad Wi-Fi environment now..

**Comment by sw005320** [comment]
Hmm, I also have the same impression. Could you make sure to double check your experiments?

**Comment by Masao-Someki** [comment]
I tried rnn/transformer model with 48k dataset again, and got the following result. 
I actually tried some parametor sets, but I don't still get good result with 48k-transformer model...
```
# 48k-rnn 
write a CER (or TER) result in exp/tr_no_dev_pytorch_train/decode_eval1_decode/result.txt
      | SPKR   | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
      | Sum/Avg|  250    5928 | 81.8   15.7    2.5    2.4   20.6   97.2 |

#48k-transformer model
write a CER (or TER) result in exp/tr_no_dev_pytorch_train/decode_eval1_decode/result.txt
      | SPKR   | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
      | Sum/Avg|  250    5928 | 75.7   17.0    7.2    2.3   26.6   98.0 |
```

**Comment by Masao-Someki** [comment]
By tuning the warmup steps, I got the best result!
```
write a CER (or TER) result in exp/tr_no_dev_pytorch_train/decode_eval1_decode/result.txt
      | SPKR   | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
      | Sum/Avg|  250    5928 | 83.1   13.6    3.3    1.8   18.7   94.8 |
```
Now I'm running the same parametor set with 16k datasets. After I get the result, I'll update RESULTS.md

**Comment by kan-bayashi** [comment]
Great! How did you tune it?

**Comment by Masao-Someki** [comment]
Using the default warmup step(25000), the accuracy curve seems like this.
image
Accuracy begins to fall from about epoch 60, so I thought warmup process from epoch 60 is not efficient.

**Comment by sw005320** [comment]
You should remove `egs/jsut/asr1/conf/pitch.conf` and `egs/jsut/asr1/conf/fbank.conf` because they are generated from `egs/jsut/asr1/local/change_sampling_rate.sh`

**Comment by sw005320** [comment]
Thanks a lot!
This is a great study.

**Pull request merged**